### DISCUSSION TITLE
I love this table: 2.0 Migration - ORM Usage.  Then, combine with AI LLM as prompt gives me coding examples

### DISCUSSION THREAD

**Discussion by taigit** [sqlalchemy/sqlalchemy#10140] [Usage Questions]
Using chat-gpt / bard to create coding examples:
One difficulty: getting v2.0 style code.

The table [2.0 Migration - ORM Usage](https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#migration-orm-usage)  is wonderful!

Moreover, the very first sentence of that table is prefixed with the perfect AI prompt, which I used to correct the LLM output.

Thus, my prompt: 

> The biggest visible change in SQLAlchemy 2.0 is the use of Session.execute() in conjunction with select() to run ORM queries, instead of using Session.query(). 
> 
> sqlalchemy orm query that is compatible with both v2.0 and v1.4: get user with username = x

and the reply:

> You are correct! In SQLAlchemy 2.0, the Session.execute() method in conjunction with select() is recommended for running ORM queries, and the Session.query() method is no longer the primary way to perform queries. 

I just mention this to say that clear, detailed sentences in the docs are used by people like me to steer chat-gpt to what we want (v2.0 style code).

_________________________________________________
ps: the code that was generated by chat-gpt

`
from sqlalchemy import create_engine, Column, String, Integer, select
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker

# Assuming you have already defined your User model using declarative_base
Base = declarative_base()

class User(Base):
    __tablename__ = 'users'
    id = Column(Integer, primary_key=True)
    username = Column(String(50), unique=True, nullable=False)
    # Add other user attributes as needed

# Replace 'your_database_uri' with the actual connection URI for your database
engine = create_engine('your_database_uri')
Base.metadata.create_all(engine)

# Create a session to interact with the database
Session = sessionmaker(bind=engine)
session = Session()

# Query for a user with a specific username
username_to_find = 'x'
stmt = select(User).where(User.username == username_to_find)
user = session.execute(stmt).scalars().first()

if user:
    # User with the given username was found
    print(f"User found: {user.username}")
else:
    # User with the given username was not found
    print("User not found.")

`

**Comment by zzzeek** [comment]
Hi -  

I'm not seeing actionable items here, so closing this.  Feel free to ask any usage questions

**Reply by taigit** [reply]
I don't know if or how I can re-open the issue.  I added a usage question in the discussion.

**Comment by taigit** [comment]
Now I have a clearer path of what I would like to do:
I'd like to leverage LLM to create 2.0-style[¶](https://docs.sqlalchemy.org/en/20/glossary.html#term-1) code.

In fact, I would like v2.0 style w/o legacy query object code and other legacy API that maintains backwards compatibility.
( version 2.0 still maintains the legacy [Query](https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query) object to allow legacy code to remain largely 2.0 compatible. )

Previously, I wanted the LLM to generate this code, but they can't be trusted too much.

New idea: use the LLM code and have sqlalchemy flag any code that is legacy or using  backwards compatible constructs.
Can this be done?


The 2.0 migration guide gives me some mechanisms:
https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#migration-to-2-0-step-four-use-the-future-flag-on-engine

Is there a sqlalchemy flag that says: not using any legacy or backwards compatible constructs ?

**Reply by CaselIT** [reply]
Hi,

I'm -1 on that, since there are no plans of removing query if that's the legacy api you are taking about.

Such functiality would probably be best integrated with something like flake8, that allows plugins to be installed for custom linting
